esp-hal/src/i2c.rs

```diff
@@ -197,29 +194,173 @@ where
+                    chunk,
+                    idx == 0,
+                    idx == chunk_count - 1,
+                    cmd_iterator,
+                )
+                .inspect_err(|_| self.internal_recover())?;
+        }
+
+        Ok(())
     }
 
     /// Writes bytes to slave with address `address` and then reads enough bytes
     /// to fill `buffer` *in a single transaction*
     pub fn write_read(
         &mut self,
         address: u8,
-        bytes: &[u8],
-        buffer: &mut [u8],
+        write_buffer: &[u8],
+        read_buffer: &mut [u8],
     ) -> Result<(), Error> {
-        self.peripheral
-            .master_write_read(address, bytes, buffer)
-            .inspect_err(|_| self.internal_recover())
+        let chunk_count = write_buffer.len().div_ceil(I2C_CHUNK_SIZE);
+        for (idx, chunk) in write_buffer.chunks(I2C_CHUNK_SIZE).enumerate() {
+            // Clear all I2C interrupts
+            self.peripheral.clear_all_interrupts();
+
+            let cmd_iterator = &mut self.peripheral.register_block().comd_iter();
+
+            self.peripheral
+                .write_operation(
+                    address,
+                    chunk,
+                    idx == 0,
+                    idx == chunk_count - 1,
+                    cmd_iterator,
+                )
+                .inspect_err(|_| self.internal_recover())?;
+        }
+
+        let chunk_count = read_buffer.len().div_ceil(I2C_CHUNK_SIZE);
+        for (idx, chunk) in read_buffer.chunks_mut(I2C_CHUNK_SIZE).enumerate() {
+            // Clear all I2C interrupts
+            self.peripheral.clear_all_interrupts();
+
+            let cmd_iterator = &mut self.peripheral.register_block().comd_iter();
+
+            self.peripheral
+                .read_operation(
+                    address,
+                    chunk,
+                    idx == 0,
+                    idx == chunk_count - 1,
+                    idx < chunk_count - 1,
+                    cmd_iterator,
+                )
+                .inspect_err(|_| self.internal_recover())?;
+        }
+
+        Ok(())
+    }
+
+    /// Execute the provided operations on the I2C bus.
+    ///
+    /// Transaction contract:
+    /// - Before executing the first operation an ST is sent automatically. This
+    ///   is followed by SAD+R/W as appropriate.
+    /// - Data from adjacent operations of the same type are sent after each
+    ///   other without an SP or SR.
+    /// - Between adjacent operations of a different type an SR and SAD+R/W is
+    ///   sent.
+    /// - After executing the last operation an SP is sent automatically.
+    /// - If the last operation is a `Read` the master does not send an
+    ///   acknowledge for the last byte.
+    ///
+    /// - `ST` = start condition
+    /// - `SAD+R/W` = slave address followed by bit 1 to indicate reading or 0
+    ///   to indicate writing
+    /// - `SR` = repeated start condition
+    /// - `SP` = stop condition
```

Comment:
In future, I hope we can make this contract optional as well and allow users to do this themselves.

I personally need a non e-hal conforming contract for the SCCB protocol.